src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -287,6 +316,84 @@ public TestResult findCorrespondingResult(String id) {
         return Collections.emptyList();
     }
 
+    /**
+     * Start for Test Result Trends.
+     * From here onwards start the code area responsible for generating various test result trends.
+     */
+
+    public hudson.tasks.junit.TestResult loadXml(){return null;}
+
+    /**
+     * A method for getting the list of packages for all levels of hierarchy.
+     * @return Array of packages of all hierarchies.
+     */
+    public String[] getProjectList(){
+
+        /*
+         * If project list is already ready no need to construct again.
+         */
+        if(projectList!=null) return projectList;
+        hudson.tasks.junit.TestResult r = loadXml();
+        Collection<SuiteResult> suiteList = r.getSuites();
+
+        /*
+         * A set for the package names.
+         */
+        Map<Integer,HashSet<String>> projectSet = new HashMap<Integer,HashSet<String>>();
```

Comment:
Use `ConcurrentHashMap`, its thread safe. It will impact performance though. But correctness is of more priority.

Reply:
Why `HashSet<String>` and not just `Set` ?